### DISCUSSION TITLE
How to get the SERIALNUMBER of a USB Device connected to my laptop on Windows OS ?

### DISCUSSION THREAD

**Discussion by abrahan-romo** [oshi/oshi#2667] [General]
I am using OSHI version 6.2.2. If I run a short program in my Eclipse IDE, it is capable of collecting the SERIAL NUMBER of a USB Camera attached to my laptop. On the other hand, if I create an independent artifact and run it in a DOS console, I am NOT able to get the SERIAL NUMBER.

How can I give OSHI credentials or authorization to collect the serial number of the USB Camera attached?

**Comment by dbwiddis** [comment]
OSHI includes serial numbers in the listing of USB Devices if they are available.

It navigates the device tree (see https://github.com/oshi/oshi/blob/master/oshi-core/src/main/java/oshi/driver/windows/DeviceTree.java)  which calls the native [SetupDiGetClassDevs](https://learn.microsoft.com/en-us/windows/win32/api/setupapi/nf-setupapi-setupdigetclassdevsw) function on Windows to list the devices and [CM_Get_DevNode_Registry_Property](https://learn.microsoft.com/en-us/windows/win32/api/cfgmgr32/nf-cfgmgr32-cm_get_devnode_registry_propertyw) to look up their properties.

There is no indication in those APIs that any special permissions are required.  All the information they fetch is also available in the registry.

How are you collecting the serial number in your own program?

**Reply by abrahan-romo** [reply]
I appreciate the feedback, thank you for that. Let me write down the steps:

1.-- create a folder in the C:\ partition. Name it as oshi.    C:\oshi
2.-- copy the following files in there: jna-5.12.1.jar, jna-platform-5.12.1.jar, oshi-core-5.7.0.jar, oshi-core-6.2.2.jar, slf4j-api-1.7.36.jar
3.-- Being in Eclipse, compile and export the source code of my oshi experiment into a jar file. The name of the Java source file is 
       OSHI_GetOsInfo.java so the class name is OSHI_GetOsInfo.class no package, and the jar file containing this class is runoshi.jar
4.-- It is assumed Java version 8 is installed.
5.-- being on the DOS console run the following instructions:
      java -classpath jna-5.12.1.jar;jna-platform-5.12.1.jar;oshi-core-5.7.0.jar;oshi-core-6.2.2.jar;slf4j-api-1.7.36.jar;runoshi.jar OSHI_GetOsInfo > OSHI.txt
6.-- There is a list of USB Devices as well as some other information. Take a look to CAMERA USB entry device, there is no SERIALNUMBER.
7.-- create a simple project, plain project use the same jars just mentioned on step 5.-- There is not need to export the OSHI_GetOsInfo.java just make right click and select run as java application copy paste the outcome execution in the Console output into a txt file named [EclipseOSHIexecution.txt](https://github.com/user-attachments/files/15894467/EclipseOSHIexecution.txt).  There is SERIALNUMBER for the CAMERA USB device.

8.-- I forgot to include the OSHI source code that I am using. including as an attachment now.....
[OSHI_GetOsInfo.txt](https://github.com/user-attachments/files/15894485/OSHI_GetOsInfo.txt)


I am including both txt files as evidence.

What Am I doing wrong in the DOS console output ?
[OSHI.txt](https://github.com/user-attachments/files/15894468/OSHI.txt)
[EclipseOSHIexecution.txt](https://github.com/user-attachments/files/15894471/EclipseOSHIexecution.txt)

Thanks !!! I appreciate the help in advance!

**Comment by dbwiddis** [comment]
Thanks for the clarification and example.  As I understand when you run OSHI from within Eclipse it performs as expected, but when you are running it from a console it does not?

I'm really not sure here, this seems to possibly be some permissions restrictions that Windows has in its Console, and not really anything associated with OSHI itself.

**Reply by abrahan-romo** [reply]
Right exactly,  you got it; that is my point. It seems to be that when OSHI is running within the Eclipse project, it has plenty of permissions and authorizations to interrogate the hardware and collect as much information as possible; in the meantime, the HW is capable of providing it. On the other hand, running the exact same portion of code in the same laptop but in a DOS console environment, there are some kind of restrictions regarding permissions and authorizations in order to collect the same information and, in particular, the SERIAL NUMBER that the HW is capable of providing it. That is why I wanted to know if there are some oshi.properties files to say something that I may use to grant access or something like that so I can finally collect the information.

**Reply by dbwiddis** [reply]
Not sure what's going on, but here's something interesting.  

Note that the camera is a composite device and the serial number is present on the parent.   You can see the serial `11S3AA06227100YK1EXK44K026` on the DeviceId of the parent composite device, which is properly parsed to populate the serial number field, and that's how it's populated on the children.
```
USB name USB Composite Device
USB vendor (Standard USB Host Controller)
USB vendor ID 0x0f66
USB product ID 0x4587
USB serial 11S3AA06227100YK1EXK44K026
USB unique DeviceId USB\VID_0F66&PID_4587\11S3AA06227100YK1EXK44K026
 
 
USB name CAMERA Camera IR
USB vendor Microsoft
USB vendor ID 0x0f66
USB product ID 0x4587
USB serial 11S3AA06227100YK1EXK44K026
USB unique DeviceId USB\VID_0F66&PID_4587&MI_02\8&209B544D&0&0002
 
 
USB name CAMERA Camera RGB
USB vendor Microsoft
USB vendor ID 0x0f66
USB product ID 0x4587
USB serial 11S3AA06227100YK1EXK44K026
USB unique DeviceId USB\VID_0F66&PID_4587&MI_00\8&209B544D&0&0000
 
 
USB name CAMERA Camera RGB
USB vendor (Generic USB Audio)
USB vendor ID 0x0f66
USB product ID 0x4587
USB serial 11S3AA06227100YK1EXK44K026
USB unique DeviceId USB\VID_0F66&PID_4587&MI_04\8&209B544D&0&0004
```
But from the command window, while the parent `11S3AA06227100YK1EXK44K026` serial is present in the Device ID but not the serial field, and is therefore not passed down:
```
USB name USB Composite Device
USB vendor (Standard USB Host Controller)
USB vendor ID 0x0f66
USB product ID 0x4587
USB serial 
USB unique DeviceId USB\VID_0F66&PID_4587\11S3AA06227100YK1EXK44K026
 
 
USB name CAMERA Camera IR
USB vendor Microsoft
USB vendor ID 0x0f66
USB product ID 0x4587
USB serial 
USB unique DeviceId USB\VID_0F66&PID_4587&MI_02\8&209B544D&0&0002
 
 
USB name CAMERA Camera RGB
USB vendor Microsoft
USB vendor ID 0x0f66
USB product ID 0x4587
USB serial 
USB unique DeviceId USB\VID_0F66&PID_4587&MI_00\8&209B544D&0&0000
 
 
USB name CAMERA Camera RGB
USB vendor (Generic USB Audio)
USB vendor ID 0x0f66
USB product ID 0x4587
USB serial 
USB unique DeviceId USB\VID_0F66&PID_4587&MI_04\8&209B544D&0&0004
```

That code is here:

https://github.com/oshi/oshi/blob/8293abcb6e2e4012e23260a8452f0381a3dd6e80/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java#L95-L112

So it seems `parseDeviceIdToVendorProductSerial()` is behaving differently in the two environments.  That method relies on this regex:
```regex
.*(?:VID|VEN)_(\p{XDigit}{4})&(?:PID|DEV)_(\p{XDigit}{4})(.*)\\(.*)
```
Comparing the two DeviceId Strings they look identical in your output text:
```
Eclipse: USB\VID_0F66&PID_4587\11S3AA06227100YK1EXK44K026
Command: USB\VID_0F66&PID_4587\11S3AA06227100YK1EXK44K026
```
Lack of serial means either the regex didn't match, or it matched and either the third capturing group was not empty or the serial contained the & character.  Neither is true here, so I am failing to see how that would cause the problem.

Some possibilities:
 - Regex handling sometimes changes between JDK versions.  I doubt that should happen here, but what JDK versions are you using in both environments?
 - Locale settings or character encoding may be different (UTF-8 vs. Windows-1252)

Next steps: Check out the actual java command line that Eclipse is using ([you can run it in debug mode to find that](https://stackoverflow.com/a/9477695/1161484)) and compare, as well as comparing your java versions.

**Reply by dbwiddis** [reply]
Other possibility here is the use of the posix regex library feature `\p{xDigit}` which might use a different regex library when run in the command window.  You could write a simple java program to try to test that as well.

**Reply by dbwiddis** [reply]
Some debug steps: try running this program in Eclipse and Command:
```
import java.util.Locale;

public class RegexTest {
    public static void main(String[] args) {
        String testString = "123abcDEF";
        String regex = "\\p{xDigit}+";
        boolean matches = testString.matches(regex);
        
        System.out.println("Matches: " + matches);
        System.out.println("File Encoding: " + System.getProperty("file.encoding"));
        System.out.println("Default Locale: " + Locale.getDefault());
        System.out.println("Java Version: " + System.getProperty("java.version"));
    }
}
```

Also try changing your command terminal codepage to UTF-8
```
chcp 65001
```

**Reply by abrahan-romo** [reply]
@dbwiddis;

Hi Daniel, My apologies in advance. I was not able to get back to you as soon as I wanted. I had to provide urgent assistance to some customers, and that activity completely took up my time.  

I have good news: It is working; I am collecting the SERIAL NUMBER that I need now. 

I got confirmation that OSHI version oshi-core-5.7.0.jar is the one that is not working correctly. On the other hand I  am using OSHI version oshi-core-6.2.2.jar and it is the one that is working just fine.

So I just did the changed version, and there you go; everything is fine now.

I really appreciate your help, your suggestions, and your time!!!! Thank you very much !!!!

**Reply by dbwiddis** [reply]
Awesome. I should have looked at the change log earlier.  This bug was introduced in 5.0.0, reported in https://github.com/oshi/oshi/issues/1522 and fixed in https://github.com/oshi/oshi/issues/1524 in version 5.5.0.
